teps/0138-improve-tekton-pipeline-feature-flags.md

```diff
@@ -0,0 +1,73 @@
+---
+status: proposed
+title: Improve Tekton Pipeline Feature Flags
+creation-date: '2023-07-07'
+last-updated: '2023-07-07'
+authors:
+- '@JeromeJu'
+- '@chitrangpatel'
+- '@lbernick'
+---
+
+# TEP-0138: Improve Tekton Pipeline Feature Flags
+
+<!-- toc -->
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+
+## Summary
+
+This document proposes updating Tekton Pipelines' feature flags design, as originally proposed in [TEP-0033](https://github.com/tektoncd/community/blob/main/teps/0033-tekton-feature-gates.md), to address problems related to the coupling API versioning and feature versioning, and to provide a better experience for cluster operators.
+
+## Motivation
+
+Currently there are two design decisions from [TEP-0033](https://github.com/tektoncd/community/blob/main/teps/0033-tekton-feature-gates.md), grouped feature flags and coupled feature version and apiVersion, that led to challenges for cluster operators, users and maintainers for the usage of existent feature flags.
+
+### Grouped feature flags
+
+New features in Tekton start at "`alpha`" stability and can be promoted to `beta` or "`stable`". Today, we have a single feature flag, `enable-api-fields`, that enables all fields in the Tekton API supporting features at a certain stability level. `enable-api-fields` can be set to "`alpha`", `beta`, or "`stable`", specifying the minimum stability level for any fields used in Tekton APIs on that cluster. In addition, features that aren't controlled by fields in Tekton APIs have their own feature flags. For more information, see [TEP-0033](https://github.com/tektoncd/community/blob/main/teps/0033-tekton-feature-gates.md). This approach has led to the following pain points: 
+
+- Cluster operators can't enable individual `alpha` or `beta` features controlled by API fields.
+  - Since features at a lower stability level tend to have more bugs, cluster operators and vendors may want to limit usage of lower-stability features they do not need, while still being able to enable individual features for specific use cases.
+  - Enabling all alpha features may have led to problems with the usage and debugging of stable and beta features. For example, in [alpha object param feature causes problems for debugging beta resolver features](https://github.com/tektoncd/pipeline/issues/6365) and [alpha propagated object params that has led to the unclear behaviour with opting in all alpha features together](https://github.com/tektoncd/pipeline/issues/5988). 
```

Comment:
I don't understand the [alpha object param feature causes problems for debugging beta resolver features](https://github.com/tektoncd/pipeline/issues/6365) link - that's not the title of issue #6365 - is that on purpose?
Sorry, I fail to understand how having individual feature flags would have helped with an issue that seems to be related to marshalling of our types in the end-user API.

Reply:
Regarding [alpha propagated object params that has led to the unclear behaviour with opting in all alpha features together](https://github.com/tektoncd/pipeline/issues/5988), I think the problem is that "propagated parameters" is **not** an API driven feature and should have been implemented as an independent behaviour feature flag.
The feature affects how we read the resource provided by the user, but it cannot be controlled by the user via some API field. I think it's apparent in the [implementation](https://github.com/tektoncd/pipeline/pull/4845/files) that the bulk of the logic is guarded by a feature flag check in the controller, instead of validation which is how we guard API driven feature flags.